### PR TITLE
Fix server side copy of files with `?` in - fixes #7058

### LINKED ISSUE #7058: Cant server side copy files with ? in the name
## Expected Behavior

I used rclone to server side copy a file with a `?` in the name.

```
$ echo "hello" | rclone rcat 'miniodev:container1/hello?hello.txt'
$ rclone ls miniodev:container1
$ rclone copyto --retries 1  -vv --dump bodies 'miniodev:container1/hello?hello.txt' 'miniodev:container1/hello2?hello2.txt'
```

I expected this to work (as it does on S3) however it produces a "NoSuchKey" error.

Here are the HTTP traces from rclone

```
2019/01/09 11:49:13 DEBUG : HTTP REQUEST (req 0xc0001bee00)
2019/01/09 11:49:13 DEBUG : PUT /container1/hello2%3Fhello2.txt HTTP/1.1
Host: 127.0.0.1:9000
User-Agent: rclone/v1.45-069-g7ee7bc87-beta
Content-Length: 0
Authorization: XXXX
Content-Type: text/plain; charset=utf-8
X-Amz-Acl: 
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Copy-Source: container1/hello2%3Fhello2.txt
X-Amz-Date: 20190109T114913Z
X-Amz-Meta-Mtime: 1547034548.472997411
X-Amz-Metadata-Directive: REPLACE
Accept-Encoding: gzip
```

and the response

```
2019/01/09 11:49:13 DEBUG : HTTP RESPONSE (req 0xc0001bee00)
2019/01/09 11:49:13 DEBUG : HTTP/1.1 404 Not Found
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Date: Wed, 09 Jan 2019 11:49:13 GMT
Server: Minio/DEVELOPMENT.GOGET (linux; amd64)
Vary: Origin
X-Amz-Bucket-Region: us-east-1
X-Amz-Request-Id: 15782BBAC878A027
X-Minio-Deployment-Id: 00ce564e-e6d3-4a99-8b80-07a7d9cfdc62
X-Xss-Protection: 1; mode=block

ef
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Resource>/container1/hello2?hello2.txt</Resource><RequestId>15782BBAC878A027</RequestId><HostId>3L137</HostId></Error>
0

```

## Possible Solution

What appears to be happening is minio is looking for the a `versionId` parameter on the end of the `x-amz-copy-source` url wheras it should be looking for it in the `x-amz-copy-source-version-id` header according to [the docs](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectCOPY.html).

I will submit a PR to fix this in just a moment :-)

## Steps to Reproduce (for bugs)

1. [Install rclone](https://rclone.org/downloads/)
2. [Configure for use with minio](https://rclone.org/s3/#minio)
3. Create a test file: `echo "hello" | rclone rcat 'miniodev:container1/hello?hello.txt'`
4. Check the file is present: `rclone ls miniodev:container1`
5. Attempt a server side copy dumping http transactions: `rclone copyto --retries 1  -vv --dump bodies 'miniodev:container1/hello?hello.txt' 'miniodev:container1/hello2?hello2.txt'`
6. Delete the destination file if it was created: `rclone delete 'miniodev:container1/hello2?hello2.txt'`

## Context

This causes the rclone integration tests to fail when run against a minio backend, indicating either rclone or minio is out of spec with s3.  In this case I think it is minio's problem since the integration tests (and that specific test) work fine with s3.

## Regression

I believe this is a regression but it is a while since I have run the rclone integration tests against minio

## Your Environment
* Version used (`minio version`): 4e6e05f8e02e4f4d85f8e2d2491f5a9af1c24752 (latest master)
* Environment name and version (e.g. nginx 1.9.1): directly run in debug
* Server type and version: linux ubuntu
* Operating System and version (`uname -a`): Linux 4.15.0-36-generic #39-Ubuntu SMP Mon Sep 24 16:19:09 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
* Link to your project: https://github.com/ncw/rclone

### PR DESCRIPTION
**Pull request by ncw**
## Description

Before this change the CopyObjectHandler and the CopyObjectPartHandler
both looked for a `versionId` parameter on the `X-Amz-Copy-Source` URL
for the version of the object to be copied.  According to S3
documentation the version should be found in the
`X-Amz-Copy-Source-Version-Id` field.

A consequence of this was that files with `?` in their names could not
be server side copied.

This fixes the problem by reading the version ID from the correct
header instead of the URL.

S3 Docs:
- https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectCOPY.html
- https://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPartCopy.html

## Motivation and Context

See #7058 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by harshavardhana** [comment]
You shouldn't remove the previous code @ncw we need to honor the legacy way of copying versionId as well which is supported by quite a few applications such as pcf backup etc https://github.com/minio/minio/issues/6935

This implementation should be in conjunction with VersionId query string and also if we cannot find VersionId query string we should honor the filenames with `?` automatically. Only special handling is required for VersionId

**Comment by ncw** [comment]
> You shouldn't remove the previous code @ncw we need to honor the legacy way of copying versionId as well which is supported by quite a few applications such as pcf backup etc #6935
> 
> This implementation should be in conjunction with VersionId query string and also if we cannot find VersionId query string we should honor the filenames with `?` automatically. Only special handling is required for VersionId

Yes this form of copying is mentioned in the examples in the s3 docs....

I'll have a go at a better fix

**Comment by ncw** [comment]
@harshavardhana I've updated the commit.

It could possibly be two commits now - let me know what you think.

File: cmd/object-handlers.go [hunk 1]
```diff
@@ -675,6 +671,21 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
...
+		// any versions other than "null".
+		if vid != "null" {
+			writeErrorResponse(w, ErrNoSuchVersion, r.URL, guessIsBrowserReq(r))
+			return
+		}
+	}
+
+	// Unescape Copy source path if necessary.
+	cpSrcPathUnescaped, err := url.QueryUnescape(cpSrcPath)
+	if err != nil {
```
Comment (harshavardhana) [inline cmd/object-handlers.go]: Shouldn't this be `if err == nil`

File: cmd/object-handlers.go [hunk 2]
```diff
@@ -1487,6 +1494,21 @@ func (api objectAPIHandlers) CopyObjectPartHandler(w http.ResponseWriter, r *htt
...
```
Comment (harshavardhana) [inline cmd/object-handlers.go]: Shouldn't this be `if err == nil` ?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/object-handlers.go (hunk 1 above)
Comment (ncw) [reply cmd/object-handlers.go]: Yes - well spotted!

File: cmd/object-handlers.go (hunk 2 above)
Comment (ncw) [reply cmd/object-handlers.go]: Yes - well spotted!

**Comment by ncw** [comment]
I will post a fix with some tests either later today or tomorrow :-)

**Comment by ncw** [comment]
I've addressed your concerns from the review and added tests for all the new code.

PTAL!

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by harshavardhana — DISMISSED** [review]
LGTM

**Comment by ncw** [comment]
I have fixed the CI so this is ready for review/merge I think.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by harshavardhana — APPROVED** [review]

**Review by donatello — APPROVED** [review]

**Comment by ncw** [comment]
Thank you :-)

**Pull request merged**